rotkehlchen/chain/evm/decoding/rainbow/decoder.py

```diff
@@ -0,0 +1,388 @@
+                if bytes_to_address(log.topics[2]) == self.router_address:
+                    sell_amount = int.from_bytes(log.data)
+
+                # Find the transfer log from the router to the swap
+                if bytes_to_address(log.topics[1]) == self.router_address:
+                    swap_amount = int.from_bytes(log.data)
+
+                if swap_amount and sell_amount:
+                    break
+
+        if not (sell_amount and swap_amount):
+            return DEFAULT_DECODING_OUTPUT
+
+        # Remove the transfer event
+        # of the purchased token to the user's wallet
+        # to create a new event with the right sequence index
+        transfer_log = all_logs[-1]
+        buy_asset = self.base.get_or_create_evm_token(transfer_log.address)
+        buy_amount = int.from_bytes(transfer_log.data)
+        remove_action = ActionItem(
+            action='skip',
+            from_event_type=HistoryEventType.RECEIVE,
+            from_event_subtype=HistoryEventSubType.NONE,
+            asset=buy_asset,
+        )
+
+        # Create a new event for receive with right index
+        tx_hash = transaction.tx_hash
+        timestamp = transaction.timestamp
+        amount = asset_normalized_value(buy_amount, self._resolve_asset(buy_asset))
+        in_event = self.base.make_event_next_index(
+            tx_hash=tx_hash,
+            timestamp=timestamp,
+            event_type=HistoryEventType.TRADE,
+            event_subtype=HistoryEventSubType.RECEIVE,
+            asset=buy_asset,
+            amount=amount,
+            location_label=sender,
+            notes=f'Receive {amount} {buy_asset.resolve_to_asset_with_symbol().symbol} as the result of a Rainbow Swap',  # noqa: E501
+            counterparty=CPT_RAINBOW_SWAPS,
+            address=self.router_address,
+        )
+        decoded_events.append(in_event)
+
+        fee_event = self._create_fee_event(
+            tx_hash=transaction.tx_hash,
+            sequence_index=in_event.sequence_index + 1,
+            timestamp=transaction.timestamp,
+            fee_raw=sell_amount - swap_amount,
+            fee_asset=out_event.asset,
+            sender=sender,
+        )
+        return DecodingOutput(event=fee_event, action_items=[remove_action])
+
+    def _maybe_decode_swap(
+        self,
+        token: EvmToken | None,  # pylint: disable=unused-argument
+        tx_log: EvmTxReceiptLog,  # pylint: disable=unused-argument
+        transaction: EvmTransaction,
+        decoded_events: list['EvmEvent'],
+        action_items: list[ActionItem],   # pylint: disable=unused-argument
+        all_logs: list[EvmTxReceiptLog],
+    ) -> DecodingOutput:
+        sender = transaction.from_address
+        first_log = all_logs[0]
+        last_log = all_logs[-1]
+        # Check if is calling the rainbow router
+        if transaction.to_address != self.router_address:
+            return DEFAULT_DECODING_OUTPUT
```

Comment:
we said that you can't trust the input data but neither you can trust the to_address since when using smart wallets it is the address of the wallet